stl/inc/generator

```diff
@@ -40,22 +40,38 @@ struct alignas(__STDCPP_DEFAULT_NEW_ALIGNMENT__) _Aligned_block {
 template <class _Alloc>
 concept _Has_real_pointers = same_as<_Alloc, void> || is_pointer_v<typename allocator_traits<_Alloc>::pointer>;
 
-template <class _Allocator>
-class _Promise_allocator { // statically specified allocator type
+template <class _Proto_allocator>
+class _Coro_promise_allocator { // statically specified allocator type
 private:
-    using _Alloc           = _Rebind_alloc_t<_Allocator, _Aligned_block>;
+    using _Alloc           = _Rebind_alloc_t<_Proto_allocator, _Aligned_block>;
     using _Alloc_size_type = allocator_traits<_Alloc>::size_type;
 
+    static size_t _Allocation_block_size(const size_t _Size) noexcept {
+        // Compute the number of _Aligned_blocks needed to store a coroutine
+        // frame of _Size bytes and, if necessary, an _Alloc.
+        if constexpr (default_initializable<_Alloc> && allocator_traits<_Alloc>::is_always_equal::value) {
+            // allocator is stateless, we need no additional space
+            return (_Size + sizeof(_Aligned_block) - 1) / sizeof(_Aligned_block);
+        } else {
+            // allocator is stateful, we need space for it
+            constexpr size_t _Align = (_STD max)(alignof(_Alloc), sizeof(_Aligned_block));
+            return (_Size + sizeof(_Alloc) + _Align - 1) / sizeof(_Aligned_block);
+        }
+    }
+
     static void* _Allocate(_Alloc _Al, const size_t _Size) {
+        // memory layout:
+        // |--- coroutine frame --|- alignment padding -|---- _Alloc -----|
+        // |--- always present ---|----------- stateful only -------------|
+        // all rounded up to a multiple of sizeof(_Aligned_block)
+
+        const size_t _Size_in_blocks = _Allocation_block_size(_Size);
         if constexpr (default_initializable<_Alloc> && allocator_traits<_Alloc>::is_always_equal::value) {
             // do not store stateless allocator
-            const size_t _Count = (_Size + sizeof(_Aligned_block) - 1) / sizeof(_Aligned_block);
-            return _Al.allocate(_Convert_size<_Alloc_size_type>(_Count));
+            return _Al.allocate(_Convert_size<_Alloc_size_type>(_Size_in_blocks));
         } else {
             // store stateful allocator
-            constexpr size_t _Align = (_STD max)(alignof(_Alloc), sizeof(_Aligned_block));
-            const size_t _Count     = (_Size + sizeof(_Alloc) + _Align - 1) / sizeof(_Aligned_block);
-            void* const _Ptr        = _Al.allocate(_Convert_size<_Alloc_size_type>(_Count));
+            void* const _Ptr = _Al.allocate(_Convert_size<_Alloc_size_type>(_Size_in_blocks));

```

Comment:
Now that `_Allocation_block_size()` has been extracted (which is an awesome clarity improvement :heart_eyes_cat:), I observe that `_Al.allocate(_Convert_size<_Alloc_size_type>(_Size_in_blocks))` is identical on both branches here. Can/should we lift out `_Ptr`? We could also push down `return _Ptr;` and invert the `if constexpr` condition so that we only do work when we need to "store stateful allocator" although I don't think that's as important for clarity.

Reply:
Also, why are we repeating `default_initializable<_Alloc> && allocator_traits<_Alloc>::is_always_equal::value` 783249 times and not giving the condition a short semantically meaningful name?